File: src/App/js/react-editor.js
import React from 'react';
import MonacoEditor from 'react-monaco-editor';
import PropTypes from 'prop-types';
import ReactResizeDetector from 'react-resize-detector';


class Editor extends React.Component {
    // editor = null;
    // monaco = null

    constructor(props) {
        super(props);
    }

    editorDidMount = (editor, monaco) => {
        this.props.editorDidMount(editor, monaco);
        this.editor = editor;
        this.monaco = monaco;

        if (this.props.eventId !== null)
            window.addEventListener(this.props.eventId, ev => {
                switch (ev.detail.eventType) {
                    case "cursorMove":
                        this.editor.setSelection(ev.detail.range);
                        this.editor.focus();
                        this.editor.revealRangeInCenter(ev.detail.range);
                        break;
                    default:
                        break;
                }
            });
    }

    onChange = (newValue, e) => {
        this.props.onChange(newValue);
    }

    onResize = () => {
        if (this.editor !== null)
            this.editor.layout();
    }

    componentDidUpdate(prevProps) {
        if (prevProps.errors !== this.props.errors) {
            this.monaco.editor.setModelMarkers(this.editor.getModel(), "FSharpErrors", this.props.errors);
            this.onResize();
        }

        if (prevProps.options !== this.props.options) {
            this.editor.updateOptions(this.props.options);
        }
    }

    render() {
        let display = this.props.isHidden ? "none" : "block";
        let className = "react-editor " + this.props.customClass;
        return (
            <div className={className} style={{ height: '100%', overflow: 'hidden', display: display }}>
                <ReactResizeDetector handleWidth handleHeight onResize={this.onResize} />
                <MonacoEditor
                    value={this.props.value}
                    options={this.props.options}
                    onChange={this.onChange}
                    editorDidMount={this.editorDidMount}
                // requireConfig={requireConfig}
                />
            </div>
        );
    }
}

function noop() { }

Editor.propTypes = {
    onChange: PropTypes.func,
    value: PropTypes.string,
    editorDidMount: PropTypes.func,
    options: PropTypes.object,
    errors: PropTypes.array,
    eventId: PropTypes.string,
    isHidden: PropTypes.bool,
    customClass: PropTypes.string
};

Editor.defaultProps = {
    onChange: noop,
    value: "",
    editorDidMount: noop,
    options: null,
    errors: [],
    eventId: null,
    isHidden: false,
    customClass: ""
};

export default Editor;

File: src/App/ReactEditor.fs
module Feliz.ReactEditor

open Fable.Core
open Fable.Core.JsInterop
open Feliz
open Browser.Types
open Browser

[<Erase>]
type editor =
    /// <summary>Triggered when the value of the Editor changed</summary>
    static member inline onChange (f : string -> unit) = Interop.mkAttr "onChange" f
    /// <summary>Set the value of the Editor</summary>
    static member inline value (value : string) = Interop.mkAttr "value" value
    /// <summary>Triggered when the Editor has been mounted</summary>
    static member inline editorDidMount (f : System.Func<Monaco.Editor.IStandaloneCodeEditor, Monaco.IExports, unit>) = Interop.mkAttr "editorDidMount" f
    /// <summary>Option to pass to the Editor</summary>
    static member inline options (options : Monaco.Editor.IEditorConstructionOptions) = Interop.mkAttr "options" options
    static member inline language (language : string) = Interop.mkAttr "language" language
    /// <summary>Errors to set into the editor
    /// <para>Currently only used for the F# editor</para>
    /// </summary>
    static member inline errors (errors : Monaco.Editor.IMarkerData[]) = Interop.mkAttr "errors" errors
    /// <summary>If specified the editor will listen to the provided event key
    /// <para>When sending the event, you need to provide an `eventType` in the detail</para>
    /// <para>Additional data might also be sent using others properties</para>
    /// </summary>
    static member inline eventId (eventId : string) = Interop.mkAttr "eventId" eventId
    /// <summary>If `true`, the editor is hidden</summary>
    static member inline isHidden (isHidden : bool) = Interop.mkAttr "isHidden" isHidden
    /// <summary>Custom class to add to the `react-editor` container</summary>
    static member inline customClass (customClass : string) = Interop.mkAttr "customClass" customClass

[<Erase>]
type ReactEditor =
    static member inline editor (properties : IReactProperty list) =
        Interop.reactApi.createElement(import "default" "./js/react-editor.jsx", createObj !!properties)


// /// Override style of the div container
// let inline Style (css: CSSProp seq): Props =
//     !!("style", keyValueList CaseRules.LowerFirst css)

// let inline editor (props: Props list) : ReactElement =
//     ofImport "default" "./js/react-editor.js" (keyValueList CaseRules.LowerFirst props) []


module Dispatch =

    let cursorMove (eventId : string) (range : Monaco.Editor.IMarkerData) =
        let data =
            jsOptions<CustomEventInit>(fun o ->
                o.detail <- createObj [
                    "eventType" ==> "cursorMove"
                    "range" ==> jsOptions<Monaco.IRange>(fun o ->
                        o.startLineNumber <- range.startLineNumber
                        o.startColumn <- range.startColumn
                        o.endLineNumber <- range.startLineNumber
                        o.endColumn <- range.startColumn
                    )
                ] |> Some
            )
        let event = CustomEvent.Create(eventId, data)
        window.dispatchEvent(event) |> ignore
